src/utils/queryOps.js

```diff
@@ -1,225 +1,416 @@
+    this.materialize();
 
-        return lodashToPairs(colName).map(([name, value]) => [
-          name,
-          selector(value),
-        ]);
-      })
-      .flatten()
-      .sortBy(([colName]) => colName)
-      .toArray();
-    const output = {};
     let i = 0;
 
-    list.forEach(row => {
-      const key = lodashFromPairs(cs.map(([name, func]) => [name, func(row)]));
-      const skey = JSON.stringify(key);
-      let triple = output[skey];
+    for (const args of this.argslist) {
+      func(...args, i);
+      i += 1;
+    }
+
+    return this;
+  }
 
-      if (!triple) {
-        triple = [[], key, i];
+  enumerate() {
+    // append extra index paramter to args
+    function* output(argslist) {
+      let i = 0;
+
+      for (const args of argslist) {
+        yield [...args, i];
         i += 1;
-        output[skey] = triple;
       }
+    }
 
-      triple[0].push(row);
-    });
+    return new Wrapper(output(this.argslist), true);
+  }
 
-    return Object.values(output);
-  },
+  args() {
+    // Convert value into args
+    return new Wrapper(this.materialize().argslist.map(arg => arg[0]));
+  }
 
-  // SELECT a.*, b.* FROM listA a LEFT JOIN listB b on b[propB]=a[propA]
-  join: function join(listA, propA, listB, propB) {
-    const lookup = frum(listB)
-      .lodashGroupBy(rowB => rowB[propB])
-      .toArray();
+  filter(func) {
+    function* output(argslist) {
+      for (const [value, ...args] of argslist)
+        if (func(value, ...args)) yield [value];
+    }
 
-    return frum(listA)
-      .map(rowA => lookup[rowA[propA]].map(rowB => ({ ...rowB, ...rowA })))
-      .flatten()
-      .toArray();
-  },
+    return new Wrapper(output(this.argslist), true);
+  }
 
-  where: function where(list, expression) {
+  where(expression) {
     // Expecting a object of {columnName: value} form to use as a filter
     // return only matching rows
     const func = row => {
-      for (const [name, value] of lodashToPairs(expression)) {
+      for (const [name, value] of Object.entries(expression)) {
         if (row[name] !== value) return false;
       }
 
       return true;
     };
 
-    return lodashFilter(list, func);
-  },
+    return this.filter(func);
+  }
 
-  exists: function exists(list, columns = null) {
+  exists(columns = null) {
     // Expect a list of column names that must exist
     let func = null;
 
-    if (columns == null) {
-      func = row => row != null;
+    if (missing(columns)) {
+      func = exists;
     } else {
-      const cols = toArray(columns);
+      const selects = toArray(columns).map(selector);
 
       func = row => {
-        for (const name of cols) {
-          const v = row[name];
-
-          if (v == null || Number.isNaN(v)) return false; // TODO: simplify
-        }
+        for (const s of selects) if (missing(s(row))) return false;
 
         return true;
       };
     }
 
-    return lodashFilter(list, func);
+    return this.filter(func);
+  }
+
+  flatten() {
```

Comment:
Can you use `flatMap` instead?

Reply:
Maybe.  The problem I am hitting up against is the libraries (lodash) and the builtin functions are not working with Iterables in general; if the inner lists are generators, then no flattening happens; everything must be an Array.  The `for of` syntax seems to work for any Iterable, while the builtin vector functions, not so much.

That said, I have not tried flatMap, so it may work to replace this `flatten()`

Reply:
Confirmed: flatMap does not work when the inner lists are iterators.